hw3/HaHaFi/multiplication.cpp

```diff
@@ -19,72 +19,59 @@ inline void Check_mutiply_valid(Matrix mat1, Matrix mat2)
         }
     }
     return ret_mat;
 }
 
-bool Check_zero_init(Matrix &mat)
-{
-    for (size_t i = 0; i < mat.n_row(); ++i)
-    {
-        for (size_t j = 0; j < mat.n_col(); ++j)
-        {
-            if (mat(i, j) != 0)
-            {
-                throw std::range_error("The matrix is not zero init");
-            }
-        }
-    }
-    return true;
-}
-
 Matrix multiply_tile(Matrix &mat1, Matrix &mat2, size_t tile_size)
 {
     Check_mutiply_valid(mat1, mat2);
-    size_t K = mat1.n_col(), M = mat1.n_row(), N = mat2.n_col();
+    const size_t K = mat1.n_col(), M = mat1.n_row(), N = mat2.n_col();
     Matrix ret_mat(M, N);
 
     for (size_t m1_start_index = 0; m1_start_index < M; m1_start_index += tile_size)
     {
-        size_t m1_end_index = std::min(m1_start_index + tile_size, M);
-        for (size_t m2_start_index = 0; m2_start_index < N; m2_start_index += tile_size)
+        const size_t m1_end_index = std::min(m1_start_index + tile_size, M);
+        for (size_t common_start_index = 0; common_start_index < K; common_start_index += tile_size)
         {
-            size_t m2_end_index = std::min(m2_start_index + tile_size, N);
-            for (size_t common_start_index = 0; common_start_index < K; common_start_index += tile_size)
+            const size_t common_end_index = std::min(common_start_index + tile_size, K);
+            for (size_t m2_start_index = 0; m2_start_index < N; m2_start_index += tile_size)
             {
-                size_t common_end_index = std::min(common_start_index + tile_size, K);
+                const size_t m2_end_index = std::min(m2_start_index + tile_size, N);
                 /*
                 start calculate multiply on  tile_size*tile_size (mat1 block * mat2 block)
                 mat1 block = (m1_start_index ~ m1_end_index) * (common_start_index ~ common_end_index)
                 mat2 block = (common_start_index ~ common_end_index) * (m2_start_index ~ m2_end_index)
                 */
                 for (size_t i = m1_start_index; i < m1_end_index; ++i)
                 {
-                    
-                    for (size_t j = m2_start_index; j < m2_end_index; ++j)
+                    const size_t leader_i = i * K;
+                    const size_t leader_ret = i * N;
+                    for (size_t _k = common_start_index; _k < common_end_index; ++_k)
                     {
-                       
-                        for (size_t _k = common_start_index; _k < common_end_index; ++_k)
+                        const size_t mat_1_index = leader_i + _k;
+                        const size_t mat_2_leader = _k * N;
+                        for (size_t j = m2_start_index; j < m2_end_index; ++j)
                         {
-                            
-                            ret_mat(i,j) += mat1(i, _k) * mat2(_k, j);
+                            ret_mat.m_buffer[leader_ret + j] += mat1.m_buffer[mat_1_index] * mat2.m_buffer[mat_2_leader + j];
```

Comment:
In the tile function, I optimize the index operation, reduce a lot of multiplication.
It result that it's faster than mkl library.

Reply:
tile function with matrix  (1000 * 1000) @ (1000 * 1000)

16x faster

Original implementation
```
multiply_tile with tile_size->64 time below:
min     max     mean    std_dev
4.765   4.771   4.770   0.003
```
Revised implementation
```
multiply_tile with tile_size->64 time below:
min     max     mean    std_dev
0.299   0.301   0.300   0.001
```